crypto/engine/eng_cnf.c

```diff
@@ -76,19 +76,34 @@ static int int_engine_configure(const char *name, const char *value, const CONF
         else if (strcmp(ctrlname, "soft_load") == 0)
             soft = 1;
         /* Load a dynamic ENGINE */
-        else if (strcmp(ctrlname, "dynamic_path") == 0) {
-            e = ENGINE_by_id("dynamic");
-            if (!e)
-                goto err;
-            if (!ENGINE_ctrl_cmd_string(e, "SO_PATH", ctrlvalue, 0))
-                goto err;
-            if (!ENGINE_ctrl_cmd_string(e, "LIST_ADD", "2", 0))
-                goto err;
-            if (!ENGINE_ctrl_cmd_string(e, "LOAD", NULL, 0))
-                goto err;
-        }
+        else if (strcmp(ctrlname, "dynamic_path") == 0)
+            dynamic_path = ctrlvalue;
```

Comment:
This change seems to be addressing something different to the raised issue, and its not clear to me that this is correct. IIUC, it defers issuing the SO_PATH, LIST_ADD and LOAD commands to the dynamic engine until we read the *next* ctrl command from config...but what if there isn't a next ctrl command?

Reply:
Would a more proper fix do the check in the dynamic_load() in eng_dyn.c?

Reply:
The reason it was deferred is that in some cases the configuration defines the engine's "engine_id" only after it defines the "dynamic_path". I understand your concern, and you are correct.

I think that now that "ID" is passed to the dynamic engine, a check inside dynamic_load() will indeed be a better option. If it sounds good to you, I'll send an updated version.

Reply:
A possible fix could be to duplicate the code responsible for the dynamic load (lines 94-104), and placing it also after the for loop in case no load occurred. A sub-function could be used for this purpose to avoid code duplication.

Does this solution sound OK? If so, I'll upload a new version for the commit according to your suggestions.

Reply:
I'd be fine with moving the check to dynamic_load().

Reply:
I'll just ask to be sure I properly understood:
* Checking the engine_id could be done inside dynamic_load() and I'll move it there
* This check requires the dynamic_engine to use the "ID" control command
* The ID command should be used only when we have the "engine_id". In ibmca the config file looks like this:
```
dynamic_path = /usr/local/lib/ibmca.so
engine_id = ibmca
```
* This means that we can't load the engine the minute we see dynamic_path, and instead the engine should be loaded once we know both "dynamic_path" and "engine_id"
* For this purpose, I'm suggesting to start the dynamic load in the else case (as is in the commit today) and also at the end of the for loop (if the else case wasn't triggered).

Does it sound OK?

Reply:
everything would be a lot easier if the engine_id and the dynamic_path statements were reversed,
like in test/recipes/90-test_gost_data/gost.conf:
```
[gost_section]
engine_id = gost
dynamic_path = $ENV::OPENSSL_GOST_ENGINE_SO
default_algorithms = ALL
CRYPT_PARAMS = id-Gost28147-89-CryptoPro-A-ParamSet
```
if that is written this way you would have the engine_id ready when dynamic_path is found.

Reply:
I agree it will be easier, but it will break existing deployments, and this is why I tried to provide a solution that will work against existing configuration files.